Some tests can't be run on filesystems that have strict utf-8 mode

Ubuntu installed on ZFS filesystem have utf8only enabled ( this can not be disabled withouth recreating filesystem )


So code such as 

```python
filename = 'r\xe9\xf1emi'.encode('latin-1')
        pathname = os.path.join(tmpdir_bytes, filename)
        with open(pathname, 'wb'):
```

would fail with something like this

```
    def test_listdir_other_encoding(self, tmpdir):                     
        """                                                                                                                                                                                                                                                                                                                     
        Some filesystems allow non-character sequences in path names.                                                                                                                                                                                                                                                           
        ``.listdir`` should still function in this case.                                                                                                                                                                                                                                                                        
        See issue #61 for details.                                                                                                                                                                                                                                                                                              
        """                                                                                                                                                                                                                                                                                                                     
        assert Path(tmpdir).listdir() == []                                                                                                                                                                                                                                                                                     
        tmpdir_bytes = str(tmpdir).encode('ascii')                                                                                                                                                                                                                                                                              
                                                                                                                                                                
        filename = 'r\xe9\xf1emi'.encode('latin-1')                                                                                                             
        pathname = os.path.join(tmpdir_bytes, filename)                                                                                                                                                                                                                                                                         
>       with open(pathname, 'wb'):                                                                                                                                                                                                                                                                                              
E       OSError: [Errno 84] Invalid or incomplete multibyte or wide character: b'/build/pytest-of-nixbld/pytest-0/test_listdir_other_encoding_Pa0/r\xe9\xf1emi'
```

Not really sure whats the good solution is, at least some people can find this it might help them,

depending on whats the purpose of that test:

 - we could either just change it to use valid utf8 filename
 - dont run that test if we find utf8 only filesystem ( dont know how you would do that)

--- a/CHANGES.rst
+++ b/CHANGES.rst
@@ -1,7 +1,9 @@
 v16.3.0
 -------
 
-Require Python 3.7 or later.
+- Require Python 3.7 or later.
+- #205: test_listdir_other_encoding now automatically skips
+  itself on file systems where it's not appropriate.
 
 v16.2.0
 -------

--- a/test_path.py
+++ b/test_path.py
@@ -547,32 +547,31 @@ class TestScratchDir:
                 with contextlib.suppress(Exception):
                     f.remove()
 
-    @pytest.mark.skipif(
-        platform.system() != "Linux",
-        reason="Only Linux allows writing invalid encodings",
-    )
-    def test_listdir_other_encoding(self, tmpdir):  # pragma: nocover
+    @pytest.fixture
+    def bytes_filename(self, tmpdir):
+        name = r'r\xe9\xf1emi'.encode('latin-1')
+        base = str(tmpdir).encode('ascii')
+        try:
+            with open(os.path.join(base, name), 'wb'):
+                pass
+        except Exception as exc:
+            raise pytest.skip(f"Invalid encodings disallowed {exc}")
+        return name
+
+    def test_listdir_other_encoding(self, tmpdir, bytes_filename):  # pragma: nocover
         """
         Some filesystems allow non-character sequences in path names.
         ``.listdir`` should still function in this case.
         See issue #61 for details.
         """
-        assert Path(tmpdir).listdir() == []
-        tmpdir_bytes = str(tmpdir).encode('ascii')
-
-        filename = 'r\xe9\xf1emi'.encode('latin-1')
-        pathname = os.path.join(tmpdir_bytes, filename)
-        with open(pathname, 'wb'):
-            pass
         # first demonstrate that os.listdir works
-        assert os.listdir(tmpdir_bytes)
+        assert os.listdir(str(tmpdir).encode('ascii'))
 
         # now try with path
         results = Path(tmpdir).listdir()
-        assert len(results) == 1
         (res,) = results
         assert isinstance(res, Path)
-        assert len(res.basename()) == len(filename)
+        assert len(res.basename()) == len(bytes_filename)
 
     def test_makedirs(self, tmpdir):
         d = Path(tmpdir)

